### PR TITLE
Fixing intellisense in vscode for src and ref projects

### PR DESCRIPTION
**Pull request by Anipik**
Related to https://github.com/dotnet/runtime/issues/33427

The intellisense in vscode was broken in vscode. This change fixes it for src and ref project.
The tests project require some more work to get it fixed.

### CODE REVIEW

File: eng/intellisense.targets [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  <!-- 
+    Trigger an error if we're unable to infer the framework identifier and version. 
+    We have to evaluate this here and not in the target because by the time the target runs,
+    Microsoft.Common.targets will have defaulted to .NETFramework,Version=v4.0
+  -->
+  <PropertyGroup Condition="'$(TargetFrameworkIdentifier)' == '' or '$(TargetFrameworkVersion)' == ''">
+    <_UnsupportedTargetFrameworkError>true</_UnsupportedTargetFrameworkError>
+  </PropertyGroup>
+
+  <Choose>
```
Comment (ericstj) [inline eng/intellisense.targets]: For every property that we derive from TargetFramework can we put it in a single props file and share that between this file and Directory.Build.props to avoid duplication?

File: eng/intellisense.targets [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  </PropertyGroup>
+
+  <!-- Versions with dots are taken as is and just given leading 'v'. -->
+  <PropertyGroup Condition="'$(TargetFrameworkVersion)' == '' and '$(_ShortFrameworkVersion)' != '' and $(_ShortFrameworkVersion.Contains('.'))">
+    <TargetFrameworkVersion>v$(_ShortFrameworkVersion)</TargetFrameworkVersion>
+  </PropertyGroup>
+
+  <!-- Versions with no dots and up to 3 characters get leading 'v' and implicit dots between characters. -->
+  <PropertyGroup Condition="'$(TargetFrameworkVersion)' == '' and '$(_ShortFrameworkVersion)' != ''">
+    <TargetFrameworkVersion Condition="$(_ShortFrameworkVersion.Length) == 1">v$(_ShortFrameworkVersion[0]).0</TargetFrameworkVersion>
```
Comment (safern) [inline eng/intellisense.targets]: Should we compare this as strings? Like:
`'$(_ShortFrameworkVersion.Length)' == '1'`. I haven't needed to do numerics comparison in MSBuild so I'm not sure.

File: src/libraries/Directory.Build.targets [hunk 3]
```diff
@@ -22,6 +22,7 @@
     <StrongNameKeyId Condition="'$(IsTestProject)' == 'true'">$(TestStrongNameKeyId)</StrongNameKeyId>
   </PropertyGroup>
 
+  <Import Project="$(RepositoryEngineeringDir)intellisense.targets" Condition="'$(IsTestProject)' != 'true'" />
```
Comment (safern) [inline src/libraries/Directory.Build.targets]: Why do we don't need this for TestProjects? Test projects can also cross target, so I guess intellisense is broken there as well.

File: src/libraries/shims/manual/Directory.Build.props [hunk 4]
```diff
@@ -14,16 +14,16 @@
     <BaseIntermediateOutputPath>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', '$(OutDirName)'))</BaseIntermediateOutputPath>
     <IntermediateOutputPath>$(BaseIntermediateOutputPath)$(TargetFramework)-$(Configuration)</IntermediateOutputPath>
     <!-- Use runtime assemblies for the TargetFramework, using TargetOS -->
-    <RefPath>$(ArtifactsBinDir)runtime/$(BuildTargetFramework)-$(TargetOS)-$(Configuration)-$(ArchGroup)/</RefPath>
+    <ShimRefPath>$(ArtifactsBinDir)runtime/$(BuildTargetFramework)-$(TargetOS)-$(Configuration)-$(ArchGroup)/</ShimRefPath>
```
Comment (safern) [inline src/libraries/shims/manual/Directory.Build.props]: Out of curiosity, why is this rename needed?

Comment (Anipik) [reply src/libraries/shims/manual/Directory.Build.props]: Earlier the refpath property was in DBP, so we can overwrite it here but after we move it DBT, it was no longer happening.  We might move these to .Targets as well. 
When we address the bigger problem of not using target framework derived properties in DBP

File: src/libraries/Directory.Build.targets (hunk 3 above)
Comment (Anipik) [reply src/libraries/Directory.Build.targets]: Yes it’s broken there as well, I am working on fixing it. But currently I have a road block there in resolving package references. I will be fixing that as well

File: eng/intellisense.targets (hunk 2 above)
Comment (Anipik) [reply eng/intellisense.targets]: This is just exact copy of the code that sdk uses to parse tfms. We need to re do it here bcoz the original sdk code gets executed before we set our tfm property.

Comment (safern) [reply eng/intellisense.targets]: Can we set the tfm property in a props/targets file that gets imported before the common targets/props? There is a property that you can set so the SDK imports your targets before theirs. 

There are 2 properties that can help do that: `CustomBeforeMicrosoftCommonTargets` and `AdditionalProjectsBeforeMicrosoftCommonTargets`. https://github.com/microsoft/msbuild/pull/2014

Comment (Anipik) [reply eng/intellisense.targets]: I tried to hook into these properties. but they are still late. TargetInference gets imported in "BeforeMicrosoftCommonTargets" and CustomBeforeMicrosoftCommonTargets targets are imported before "MicrosoftCommonTargets"

Comment (safern) [reply eng/intellisense.targets]: Hmm, that's unfortunate. @ericstj any suggestions for that?

Comment (ericstj) [reply eng/intellisense.targets]: We already looked at the hooks and couldn't see anywhere to get in targets that are before the SDK parsing, which is why we're duplicating it.  Here's a sample PP from a clean new project that illustrates this: https://gist.github.com/ericstj/c213e1243e5558d1346d84034920abea#file-out-pp-L654-L900

Comment (safern) [reply eng/intellisense.targets]: Ok, that's unfortunate 😢

Comment (ericstj) [reply eng/intellisense.targets]: Yeah, we'll talk about this with SDK / MSBuild / Nuget to see if we can get a better story in the future.

File: src/libraries/dir.common.props [hunk 5]
```diff
@@ -0,0 +1,62 @@
+<Project>
```
Comment (safern) [inline src/libraries/dir.common.props]: maybe, targetframework.props?

File: src/libraries/dir.common.props [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      </PropertyGroup>
+    </When>
+  </Choose>
+
+  <PropertyGroup>
+    <TargetsNetCoreApp Condition="$(TargetFramework.StartsWith('netcoreapp'))">true</TargetsNetCoreApp>
+    <TargetsNetStandard Condition="$(TargetFramework.StartsWith('netstandard'))">true</TargetsNetStandard>
+    <TargetsNetFx Condition="$(TargetFramework.StartsWith('net4'))">true</TargetsNetFx>   
+  </PropertyGroup>
+</Project>
```
Comment (safern) [inline src/libraries/dir.common.props]: Nit: new line at end of file.

File: src/libraries/intellisense.targets [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  </PropertyGroup>
+
+  <Import Project="$(MSBuildThisDirectory)targetframework.props" />
+  
+  <PropertyGroup>
+    <RefPath>$([MSBuild]::NormalizeDirectory('$(RefRootPath)', '$(TargetFramework)'))</RefPath>
+    <BaseIntermediateOutputPath>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', '$(MSBuildProjectName)'))</BaseIntermediateOutputPath>
+    <IntermediateOutputPath>$(BaseIntermediateOutputPath)$(TargetFramework)-$(TargetFrameworkSuffix)-$(Configuration)\</IntermediateOutputPath>
+    <IntermediateOutputPath Condition="'$(TargetFrameworkSuffix)' == ''">$(BaseIntermediateOutputPath)$(TargetFramework)-$(Configuration)\</IntermediateOutputPath>
+  </PropertyGroup>
```
Comment (ericstj) [inline src/libraries/intellisense.targets]: These are also duplicated with Directory.Build.props, why didn't you move to shared props file?

File: src/libraries/intellisense.targets [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  
+  <PropertyGroup Condition="'$(TargetFrameworkIdentifier)' != '' and '$(TargetFrameworkVersion)' != ''">
+    <TargetFrameworkMoniker Condition="'$(TargetFrameworkMoniker)' == '' and '$([System.String]::IsNullOrWhitespace($(TargetFrameworkProfile)))' != 'true'">$(TargetFrameworkIdentifier),Version=$(TargetFrameworkVersion),Profile=$(TargetFrameworkProfile)</TargetFrameworkMoniker>
+    <TargetFrameworkMoniker Condition="'$(TargetFrameworkMoniker)' == ''">$(TargetFrameworkIdentifier),Version=$(TargetFrameworkVersion)</TargetFrameworkMoniker>
+  </PropertyGroup>
+  
+  <PropertyGroup>
+    <ProjectAssetsFile Condition="'$(ProjectAssetsFile)' == ''">$(MSBuildProjectExtensionsPath)/project.assets.json</ProjectAssetsFile>
```
Comment (ericstj) [inline src/libraries/intellisense.targets]: Why is this here?  I don't see MSBuildProjectExtensionsPath changed anywhere above?

File: src/libraries/intellisense.targets [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                                            '$(TargetFrameworkIdentifier)' == '.NETStandard'">false</CopyLocalLockFileAssemblies>
+
+    <!-- Don't copy local for netcoreapp projects before 3.0 or non-exe and non-component projects. -->
+    <CopyLocalLockFileAssemblies Condition="'$(CopyLocalLockFileAssemblies)' == '' and
+                                            '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                                            ('$(_TargetFrameworkVersionWithoutV)' &lt; '3.0' or
+                                             ('$(HasRuntimeOutput)' != 'true' and '$(EnableDynamicLoading)' != 'true'))">false</CopyLocalLockFileAssemblies>
+
+    <!-- All other project types should copy local. -->
+    <CopyLocalLockFileAssemblies Condition="'$(CopyLocalLockFileAssemblies)' == ''">true</CopyLocalLockFileAssemblies>
```
Comment (ericstj) [inline src/libraries/intellisense.targets]: CopyLocalLockFileAssemblies shouldn't be relevent for design-time build.  This is only for full-build about wether it decides to copy the libs from nuget packages.  You can omit all these.

File: src/libraries/intellisense.targets [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    <ContentPreprocessorOutputDirectory Condition="'$(ContentPreprocessorOutputDirectory)' == ''">$(IntermediateOutputPath)NuGet\</ContentPreprocessorOutputDirectory>
+
+    <UseTargetPlatformAsNuGetTargetMoniker Condition="'$(UseTargetPlatformAsNuGetTargetMoniker)' == '' AND '$(TargetFrameworkMoniker)' == '.NETCore,Version=v5.0'">true</UseTargetPlatformAsNuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(NuGetTargetMoniker)' == '' AND '$(UseTargetPlatformAsNuGetTargetMoniker)' == 'true'">$(TargetPlatformIdentifier),Version=v$([System.Version]::Parse('$(TargetPlatformMinVersion)').ToString(3))</NuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(NuGetTargetMoniker)' == '' AND '$(UseTargetPlatformAsNuGetTargetMoniker)' != 'true'">$(TargetFrameworkMoniker)</NuGetTargetMoniker>
```
Comment (ericstj) [inline src/libraries/intellisense.targets]: `UseTargetPlatformAsNuGetTargetMoniker ` is only for handling UAP.  Let's just remove all these.  We can just set `<NuGetTargetMoniker>$(TargetFrameworkMoniker)</NuGetTargetMoniker>`

File: src/libraries/intellisense.targets (hunk 7 above)
Comment (ericstj) [reply src/libraries/intellisense.targets]: BaseIntermediateOutputPath doesn't seem to need redefining.

Just a thought here: what if we chose a convention for IntermediateOutputPath such that it evaluated to the same thing regardless of if the TFM stripping happened or not?

Comment (Anipik) [reply src/libraries/intellisense.targets]: Currently tfm in runtime repo refers to just tfm without the rid. we can expose a property which tfm-rid if rid is present and tfm otherwise.

**Comment by Anipik** [comment]
@ericstj @safern @ViktorHofer can you take a look here ? this one is ready for review

File: src/libraries/Directory.Build.props [hunk 11]
```diff
@@ -314,9 +239,7 @@
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
 
     <RuntimePath Condition="'$(RuntimePath)' == ''">$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'runtime', '$(BuildSettings)'))</RuntimePath>
-    <RefRootPath>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'ref'))</RefRootPath>
```
Comment (ericstj) [inline src/libraries/Directory.Build.props]: This should remain. It’s used below

File: src/libraries/Directory.Build.props [hunk 12]
```diff
@@ -348,9 +271,6 @@
     <!-- setting the output paths -->
     <OutputPath>$(BaseOutputPath)$(TargetFramework)-$(TargetFrameworkSuffix)-$(Configuration)\</OutputPath>
     <OutputPath Condition="'$(TargetFrameworkSuffix)' == ''">$(BaseOutputPath)$(TargetFramework)-$(Configuration)\</OutputPath>
-    <BaseIntermediateOutputPath>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', '$(MSBuildProjectName)'))</BaseIntermediateOutputPath>
```
Comment (ericstj) [inline src/libraries/Directory.Build.props]: This should remain I think since it doesn’t need TargetFramework.

File: src/libraries/intellisense.targets [hunk 13]
```diff
@@ -0,0 +1,76 @@
+<Project>
```
Comment (ericstj) [inline src/libraries/intellisense.targets]: As I mentioned previously this needs a comment explaining why we need this and a link to where we copied from

File: src/libraries/Directory.Build.props (hunk 11 above)
Comment (Anipik) [reply src/libraries/Directory.Build.props]: This needs to be moved bcoz <RefPath> depends on this and we need to import the targetFramework.props before this block of path properties.
If we want to keep <RefRootPath> in this file we need to move this property to just before the import.

File: src/libraries/Directory.Build.props (hunk 12 above)
Comment (Anipik) [reply src/libraries/Directory.Build.props]: The reason for move is same as https://github.com/dotnet/runtime/pull/34025#discussion_r400626301

It will be nice to keep all IntermediateOutput path properties together.

File: src/libraries/Directory.Build.props (hunk 11 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props]: Don't we need to redefine `BuildTargetFrameworkRefPath` and other properties which rely on `RefRootPath` or transitives?

Comment (Anipik) [reply src/libraries/Directory.Build.props]: RefRootath is same for all tfms and will be defined before we evaluate BuildTargetFrameworkRefPath.

**Review by ericstj — APPROVED** [review]
